packages/shorebird_cli/lib/src/cache.dart

```diff
@@ -0,0 +1,126 @@
+import 'dart:io' hide Platform;
+import 'dart:isolate';
+
+import 'package:archive/archive_io.dart';
+import 'package:http/http.dart' as http;
+import 'package:path/path.dart' as p;
+import 'package:platform/platform.dart';
+import 'package:shorebird_cli/src/engine_revision.dart';
+import 'package:shorebird_cli/src/shorebird_environment.dart';
+
+class Cache {
+  Cache({
+    http.Client? httpClient,
+    this.extractArchive = extractArchiveToDisk,
+    Platform platform = const LocalPlatform(),
+  }) : httpClient = httpClient ?? http.Client() {
+    registerArtifact(PatchArtifact(cache: this, platform: platform));
+  }
+
+  final http.Client httpClient;
+  final ArchiveExtracter extractArchive;
+
+  void registerArtifact(CachedArtifact artifact) => _artifacts.add(artifact);
+
+  Future<void> updateAll() async {
+    for (final artifact in _artifacts) {
+      if (await artifact.isUpToDate()) {
+        continue;
+      }
+
+      await artifact.update();
+    }
+  }
+
+  /// Get a named directory from with the cache's artifact directory;
+  /// for example, `foo` would return `bin/cache/artifacts/foo`.
+  Directory getArtifactDirectory(String name) {
+    return Directory(p.join(shorebirdArtifactsDirectory.path, name));
+  }
+
+  /// The Shorebird cache directory.
+  static Directory get shorebirdCacheDirectory => Directory(
+        p.join(ShorebirdEnvironment.shorebirdRoot.path, 'bin', 'cache'),
+      );
+
+  /// The Shorebird cached artifacts directory.
+  static Directory get shorebirdArtifactsDirectory => Directory(
+        p.join(shorebirdCacheDirectory.path, 'artifacts'),
+      );
+
+  final List<CachedArtifact> _artifacts = [];
+
+  String get storageBaseUrl => 'https://storage.googleapis.com';
+
+  String get storageBucket => 'download.shorebird.dev';
+}
+
+typedef ArchiveExtracter = void Function(Archive, String);
+
+abstract class CachedArtifact {
+  CachedArtifact({required this.cache});
+
+  final Cache cache;
+
+  String get name;
+
+  String get storagePath;
+
+  List<String> get executables => [];
+
+  Directory get location => cache.getArtifactDirectory(name);
+
+  Future<bool> isUpToDate() async => location.existsSync();
+
+  Future<void> update() async {
+    final url = '${cache.storageBaseUrl}/${cache.storageBucket}/$storagePath';
+    final request = http.Request('GET', Uri.parse(url));
+    final response = await cache.httpClient.send(request);
+    final tempDir = Directory.systemTemp.createTempSync();
+    final archivePath = p.join(tempDir.path, '$name.zip');
+    final outputPath = location.path;
+    await response.stream.pipe(File(archivePath).openWrite());
+
+    await Isolate.run(
+      () async {
+        final inputStream = InputFileStream(archivePath);
+        final archive = ZipDecoder().decodeBuffer(inputStream);
+        cache.extractArchive(archive, outputPath);
+      },
+    );
+
+    for (final executable in executables) {
+      final process = await Process.start(
```

Comment:
I suspect this won't work on windows, but that can also be a next-week problem.

Reply:
yeah good catch

Reply:
fixed 👍